R/join_keys.R

```diff
@@ -0,0 +1,651 @@
+#' it will return the `join_keys` of that object.
+#' When called with 1 or more `join_key_set` it will create a new `join_keys` with
+#' constructed from the arguments.
+#'
+#' @return `join_keys` object.
+#'
+#' @export
+#'
+#' @examples
+#' # Setting join keys ----
+#'
+#' jk <- join_keys(
+#'   join_key("dataset_A", "dataset_B", c("col_1" = "col_a")),
+#'   join_key("dataset_A", "dataset_C", c("col_2" = "col_x", "col_3" = "col_y"))
+#' )
+#' jk
+#'
+#' # or
+#' jk <- join_keys()
+#' jk[["dataset_A"]][["dataset_B"]] <- c("col_1" = "col_a")
+#' jk[["dataset_A"]][["dataset_C"]] <- c("col_2" = "col_x", "col_3" = "col_y")
+#' jk
+#'
+#' td <- teal_data(join_keys = join_keys(join_key("a", "b", "c")))
+#' join_keys(td)
+#'
+#' jk <- join_keys()
+#' join_keys(jk)
+#'
+#' jk <- join_keys()
+#' jk <- c(jk, join_keys(join_key("a", "b", "c")))
+#' jk <- c(jk, join_keys(join_key("a", "b", "c"), join_key("a", "b2", "c")))
+join_keys <- function(...) {
+  if (missing(...)) {
+    return(new_join_keys())
+  }
+  x <- rlang::list2(...)
+  if (length(x) > 1) {
+    return(join_keys.default(...))
+  }
+  UseMethod("join_keys", x[[1]])
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.join_keys <- function(...) {
+  x <- rlang::list2(...)
+  x[[1]]
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.teal_data <- function(...) {
+  x <- rlang::list2(...)
+  x[[1]]@join_keys
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.TealData <- function(...) {
+  x <- rlang::list2(...)
+  x[[1]]$get_join_keys()
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.default <- function(...) {
+  x <- rlang::list2(...)
+  # Constructor
+  res <- new_join_keys()
+  if (length(x) > 0) {
+    join_keys(res) <- x
+  }
+
+  res
+}
+
+#' @rdname join_keys
+#' @details
+#' The setter assignment `join_keys(obj) <- ...` will merge obj and `...` if obj
+#' is not empty.
+#'
+#' @param join_keys_obj (`join_keys`) empty object to set the new relationship pairs.
+#' @param value (`join_key_set` or list of `join_key_set`) relationship pairs to add
+#' to `join_keys` list.
+#'
+#' @export
+`join_keys<-` <- function(join_keys_obj, value) {
+  UseMethod("join_keys<-", join_keys_obj)
+}
+
+#' @rdname join_keys
+#' @export
+#' @examples
+#'
+#' # Using the setter (assignment) ----
+#'
+#' jk <- join_keys()
+#' join_keys(jk) <- join_key("ds1", "ds2", "some_col")
+#' join_keys(jk) <- c(join_keys(jk), join_key("ds3", "ds4", "some_col2"))
+#'
+#' join_keys(jk)[["ds1"]][["ds3"]] <- "some_col3"
+#' jk
+`join_keys<-.join_keys` <- function(join_keys_obj, value) {
+  if (missing(value)) {
+    return(join_keys_obj)
+  }
+
+  # Assume assignment of join keys as a merge operation
+  #  Needed to support join_keys(jk)[ds1, ds2] <- "key"
+  if (checkmate::test_class(value, classes = c("join_keys", "list"))) {
+    return(value)
+  }
+
+  if (inherits(value, "join_key_set")) value <- list(value)
+
+  checkmate::assert_list(value, types = "join_key_set", min.len = 1)
```

Comment:
Is this assertion useful? How about placing argument checks on `value` in the generic?

Reply:
Good point, changed it to a multi-assertion to help users around.

Allowed to set an empty list. It seems like an odd thing to not allow.

Done in 6746e5f9